slamd/discovery/processing/algorithms/plot_generator.py

```diff
@@ -1,30 +1,52 @@
 import json
+import pandas as pd
 import plotly
 import plotly.express as px
+from sklearn.manifold import TSNE
 
 
 class PlotGenerator:
 
     @classmethod
-    def create_target_scatter_plot(cls, target_list):
-        dimensions = [col for col in target_list.columns if col != 'Utility']
+    def create_target_scatter_plot(cls, plot_df):
+        dimensions = [col for col in plot_df.columns if col != 'Utility' and col != 'Row number']
         if len(dimensions) == 1:
-            # Generate a simple scatter plot if there is only one target property
-            fig = px.scatter(target_list, x=dimensions[0], y='Utility')
-            fig.update_traces(hovertemplate='X: %{x:.2f}, Y: %{y:.2f}')
+            # Generate a simple scatter plot if there is only one target property.
+            # We include the Utility color-coded for aesthetic reasons.
+            fig = px.scatter(plot_df, x=dimensions[0], y='Utility', color='Utility',
+                             custom_data=['Row number'], title='Scatter plot of target properties')
         elif len(dimensions) == 2:
             # Plotly 5.10 issue: px.scatter_matrix() does not output anything when the matrix is 1x1.
             # We need to handle this case separately and generate a single scatter plot.
-            fig = px.scatter(target_list, x=dimensions[0], y=dimensions[1],
-                             color='Utility', title='Scatter plot of target properties')
-            fig.update_traces(hovertemplate='X: %{x:.2f}, Y: %{y:.2f}, Utility: %{marker.color:.2f}')
+            fig = px.scatter(plot_df, x=dimensions[0], y=dimensions[1], color='Utility',
+                             custom_data=['Row number'], title='Scatter plot of target properties')
         else:
             # General case
-            fig = px.scatter_matrix(target_list, dimensions=dimensions, color='Utility',
-                                    title='Scatter matrix of target properties')
+            fig = px.scatter_matrix(plot_df, dimensions=dimensions, color='Utility',
+                                    custom_data=['Row number'], title='Scatter matrix of target properties')
             # Format the tooltips in a generic way good enough for all subplots
-            fig.update_traces(diagonal_visible=False, showupperhalf=False,
-                              hovertemplate='X: %{x:.2f}, Y: %{y:.2f}, Utility: %{marker.color:.2f}')
+            fig.update_traces(diagonal_visible=False, showupperhalf=False)
 
+        fig.update_traces(
+            hovertemplate='Row number: %{customdata}, X: %{x:.2f}, Y: %{y:.2f}, Utility: %{marker.color:.2f}'
+        )
+        fig.update_layout(height=1000)
+        return json.dumps(fig, cls=plotly.utils.PlotlyJSONEncoder)
+
+    @classmethod
+    def create_tsne_input_space_plot(cls, features_df, candidate_or_target):
```

Comment:
Discuss whether the TSNE plot should be shown automatically to fix the width issue